File: ProgramingLanguage/Java/LeetCode10.java
__________________________________________________________________________________________________
1ms
class Solution {
    public boolean isMatch(String s, String p) {
        Boolean[][] memo = new Boolean[s.length() + 1][p.length() + 1];
        return dp(s, p, 0, 0, memo);
    }
    
    private boolean dp(String s, String p, int i, int j, Boolean[][] memo) {
        if (memo[i][j] != null) return memo[i][j];
        boolean ans;
        if (j == p.length()) {
            ans = i == s.length();
        } else {
            boolean firstMatch = i < s.length() && (s.charAt(i) == p.charAt(j) || p.charAt(j) == '.');
            if (j + 1 < p.length() && p.charAt(j + 1) == '*') {
                ans = dp(s, p, i, j + 2, memo) || (firstMatch && dp(s, p, i + 1, j, memo));
            } else {
                ans = firstMatch && dp(s, p, i + 1, j + 1, memo);
            }
        }
        memo[i][j] = ans;
        return ans;
    }
}
__________________________________________________________________________________________________
2ms
class Solution {
    public boolean isMatch(String s, String p) {
        boolean[][] match = new boolean[s.length()+1][p.length()+1];
        match[0][0] = true;
        for ( int j = 2; j <= p.length(); j++)
            match[0][j] = match[0][j-2] && p.charAt(j-1) == '*';
        
        for ( int i = 1; i <= s.length(); i++)
            for( int j = 1; j <= p.length(); j++){
                if (p.charAt(j-1) == '*')
                match[i][j] = match[i-1][j] && (s.charAt(i-1) == p.charAt(j-2) || p.charAt(j-2) == '.') || 
                              match[i][j-2] && p.charAt(j-1) == '*';
                else 
                    match[i][j] = match[i-1][j-1] && (s.charAt(i-1) == p.charAt(j-1) || p.charAt(j-1) == '.');
            }
        
        return match[s.length()][p.length()];
    }
}
__________________________________________________________________________________________________
3ms
class Solution {
    public boolean isMatch(String s, String p) {
        boolean[][] dp = new boolean[s.length() + 1][p.length() + 1];
        dp[s.length()][p.length()] = true;
        
        for (int i = s.length(); i >= 0; i--) {
            for (int j = p.length() - 1; j >= 0; j--) {
                boolean firstMatch = (i < s.length()) && (s.charAt(i) == p.charAt(j) || p.charAt(j) == '.');
                if (j + 1 < p.length() && p.charAt(j + 1) == '*') {
                    dp[i][j] = dp[i][j + 2] || (firstMatch && dp[i + 1][j]);
                } else {
                    dp[i][j] = firstMatch && dp[i + 1][j + 1];
                }
            }
        }
        return dp[0][0];
    }
}
__________________________________________________________________________________________________
35452 kb
class Solution {    
    char[] ct, cp;
    int lt, lp;

    private boolean solve(int it, int ip) {
        if (ip == lp) return it == lt;
        boolean fm = it < lt && (ct[it] == cp[ip] || cp[ip] == '.');
        if (ip + 1 < lp && cp[ip + 1] == '*')
            return fm && solve(it + 1, ip) || solve(it, ip + 2);
        else return fm && solve(it + 1, ip + 1);
    }
    
    public boolean isMatch(String t, String p) {
        lt = t.length(); lp = p.length();
        cp = p.toCharArray(); ct = t.toCharArray();
        return solve(0, 0);
    }
}
__________________________________________________________________________________________________
35524 kb
class Solution {
    public boolean isMatch(String s, String p) {
        int M = s.length(), N = p.length();
        boolean[][] dp = new boolean[M+1][N+1];
        dp[0][0] = true;
        
        for (int i = 0; i <= M; i++) {
            for (int j = 0; j <= N; j++) {
                if (j == 0 && i > 0) {
                    dp[i][0] = false;
                } else {
                    if (i > 0 && j > 0 && (s.charAt(i-1)==p.charAt(j-1)
                        || p.charAt(j-1) == '.')) {
                        dp[i][j] = dp[i-1][j-1];
                    } else if (j >0 && p.charAt(j-1) == '*') {
                        if (j > 1) {
                            dp[i][j] = dp[i][j-2];
                            if (i > 0 && (s.charAt(i-1) == p.charAt(j-2) 
                                          || p.charAt(j-2) == '.')){
                                dp[i][j] = dp[i][j] || dp[i-1][j];
                            }
                        }
                    }
                }
            }
        }
        return dp[M][N];
    }
}
__________________________________________________________________________________________________

File: ProgramingLanguage/C++/LeetCode10.cpp
__________________________________________________________________________________________________
4ms
class Solution {
public:
    bool isMatch(string s, string p) {
        const int m = s.size(), n = p.size();
        int i = 0, j = 0;
        bool dp[n+1][m+1];                                       // i - n - p, j - m - s
        memset(dp, false, sizeof(bool)*(m+1)*(n+1));             // i at string corresponds with i+1 at dp matrix
        dp[0][0] = true;
        while (j++ < n){                                         // scan * while i = 0
            if (j > 1 && p[j-1] == '*') dp[j][0] = dp[j-2][0];
        }
        while (i++ < m){
            j = 0;
            while (j++ < n){
                if (p[j-1] == s[i-1] || p[j-1] == '.') dp[j][i] = dp[j-1][i-1];       // if not * but match
                else if (p[j-1] == '*'){                                              // if * |
                    if (p[j-2] != s[i-1] && p[j-2] != '.') dp[j][i] = dp[j-2][i];     //      | not match
                    else dp[j][i] = (dp[j-2][i] || dp[j-1][i] || dp[j][i-1]);         //      | match
                }
            }
        }
        return dp[n][m];
    }
};
__________________________________________________________________________________________________
8ms
class Solution {
public:
    bool isMatch(string s, string p) {
        int m = s.size(), n = p.size();
        vector<vector<bool>> dp(m+1, vector<bool>(n+1, false));
        dp[0][0] = true;
        for(int i=0; i<=m; ++i)
        {
            for(int j=1; j<=n; ++j)
            {
                if(p[j-1] == '*')
                {
                    if(j>=2 && dp[i][j-2])
                        dp[i][j] = true; // repeat 0 
                    if(j>=2 && i>=1)
                        if(p[j-2] == s[i-1] || p[j-2] == '.')
                            if(dp[i-1][j] == true)
                                dp[i][j] = true;
                }
                else 
                {
                    if(i>=1 && (s[i-1]==p[j-1] || p[j-1] == '.')) 
                        dp[i][j] = dp[i-1][j-1];
                }
            }
        }
        return dp[m][n];
    }
};
__________________________________________________________________________________________________
12ms
class Solution {
public:
    bool isMatch(string s, string p) 
    {
                
        dp.resize(s.size() + 1);
        for(auto& it : dp)
        {
            it.resize(p.size(), -1);
        }
                
        return helper(s, 0, p, 0);
        
    }
private:    
    vector<vector<int>> dp;
    int real_end;
    
    bool helper(const string& S, int si, const string& P, int pi)
    {        
        if(si > S.size())
        {
            return pi >= P.size();
        }
        
        if(pi >= P.size())
        {
            return si >= S.size();
        }
        
        if(dp[si][pi] != -1)
        {
            return dp[si][pi] == 1;
        }
        
        char s = si < S.size() ? S[si] : 0;
        char p = P[pi];
        
        bool ret = false;
        if(s != 0 && (s == p || p == '.'))
        {
            ret = helper(S, si + 1, P, pi + 1);
        }                
        else if(p == '*')
        {
            char prev = pi > 0 ? P[pi-1] : 0;
            if(prev == 0)
            {
                ret = false;
            }
            else
            {
                ret =  helper(S, si, P, pi + 1);
                if(prev == '.' || prev == s)
                {
                    ret |= helper(S, si + 1, P, pi);         
                }
            }            
        }
        
        if(pi < P.size() - 1 && P[pi + 1] == '*')
        {
            ret |= helper(S, si, P, pi + 2);
        }
        
        dp[si][pi] =  ret ? 1 : 0;
        return ret;
        
    }
};
__________________________________________________________________________________________________
8228 kb
static int pre = []() {
	ios::sync_with_stdio(false);
	cin.tie(nullptr);
	return 0;
}();

class Solution {
public:
	bool isMatchRecursive(const string& s, const string& p, int strIndex, int patternIndex) {
		while (strIndex < s.size() && patternIndex < static_cast<int>(p.size() - 1)) {
			if (p[patternIndex + 1] != '*') {
				if (p[patternIndex] != '.') {
					if (s[strIndex] == p[patternIndex]) {
						strIndex++;
						patternIndex++;
					}
					else {
						return false;
					}
				}
				else {
					strIndex++;
					patternIndex++;
				}
			}
			else {
				if (p[patternIndex] != '.') {
					while (true) {
						bool success = isMatchRecursive(s, p, strIndex, patternIndex + 2);
						if (success)
							return true;
						else if (strIndex < s.size() && s[strIndex] == p[patternIndex])
							strIndex++;
						else
							break;
					}
					patternIndex += 2;
				}
				else {
					do {
						bool success = isMatchRecursive(s, p, strIndex, patternIndex + 2);
						if (success)
							return true;
						else
							strIndex++;
					} while (strIndex < s.size());
					patternIndex += 2;
				}
			}
		}
		if (strIndex < s.size() && patternIndex == p.size() - 1) {
			if (p[patternIndex] != '.') {
				if (s[strIndex] == p[patternIndex]) {
					strIndex++;
					patternIndex++;
				}
				else {
					return false;
				}
			}
			else {
				strIndex++;
				patternIndex++;
			}
		}
		if (patternIndex >= p.size()) {
			if (strIndex >= s.size())
				return true;
			else
				return false;
		}
		else {
			while (patternIndex < (int)p.size() - 1) {
				if (p[patternIndex + 1] != '*') {
					return false;
				}
				else {
					patternIndex += 2;
				}
			}
			if (patternIndex == p.size() - 1) {
				return false;
			}
			return true;
		}

	}
	bool isMatch(string s, string p) {
		return isMatchRecursive(s, p, 0, 0);
	}
};
__________________________________________________________________________________________________
9084 kb 
static int x = [](){
    std::ios::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);
    return NULL;
}();

class Solution {
public:
    bool isMatch(string s, string p) {
        int s_l = s.length(), p_l = p.length();
        int maxn = 1000+10;
        int d[maxn][maxn];
        for(int i = 0; i<=s_l; i++){
            memset(d[i], 0, sizeof(int)*maxn);
        }
        //d[i][j]: s[i-1] and p[j-1] match
        d[0][0] = 1;
        for(int i = 0; i<=s_l; i++){
            for(int j = 0; j<=p_l; j++){
                if(j>=1 && d[i][j-1] && p[j-1]=='*') d[i][j] = 1;
                if(i>=1 && j>=2 && d[i-1][j] && p[j-1] == '*' && (s[i-1] == p[j-2] || p[j-2] == '.')) d[i][j] = 1;
                if(i>=1 && j>=1 && d[i-1][j-1] &&  (s[i-1]==p[j-1] || p[j-1] == '.')) d[i][j] = 1;
                if(j>=2 && d[i][j-2] && p[j-1] == '*') d[i][j] =1;
            }
        }
        return (bool)d[s_l][p_l];
    }
};
__________________________________________________________________________________________________
